packages/rum-recorder/src/domain/segmentCollection.ts

```diff
@@ -0,0 +1,161 @@
+// concatenating multiple segments together. Segments have a size overhead (meta), so our goal is to
+// build segments containing as much records as possible while complying with the various flush
+// strategies to guarantee a good replay quality.
+//
+// When the recording starts, a segment is initially created.  The segment is flushed (finalized and
+// sent) based on various events (non-exhaustive list):
+//
+// * the page visibility change or becomes to unload
+// * the segment duration reaches a limit
+// * the encoded segment size reaches a limit
+// * ...
+//
+// A segment cannot be created without its context.  If the RUM session ends and no session id is
+// available when creating a new segment, records will be ignored, until the session is renewed and
+// a new session id is available.
+//
+// Empty segments (segments with no record) aren't useful and should be ignored.
+//
+// To help investigate session replays issues, each segment is created with a "creation reason",
+// indicating why the session has been created.
+
+export function startSegmentCollection(
+  lifeCycle: LifeCycle,
+  applicationId: string,
+  session: RumSession,
+  parentContexts: ParentContexts,
+  send: (data: Uint8Array, meta: SegmentMeta) => void
+) {
+  const worker = createDeflateWorker()
+  return doStartSegmentCollection(
+    lifeCycle,
+    () => doGetSegmentContext(applicationId, session, parentContexts),
+    send,
+    worker
+  )
+}
+
+export function doStartSegmentCollection(
+  lifeCycle: LifeCycle,
+  getSegmentContext: () => SegmentContext | undefined,
+  send: (data: Uint8Array, meta: SegmentMeta) => void,
+  worker: DeflateWorker,
+  emitter: EventEmitter = window
+) {
+  let currentSegment: Segment | undefined
+  let currentSegmentExpirationTimeoutId: ReturnType<typeof setTimeout>
+  let nextSegmentCreationReason: CreationReason = 'init'
+
+  const writer = new DeflateSegmentWriter(
+    worker,
+    (size) => {
+      if (size > SEND_BEACON_BYTE_LENGTH_LIMIT) {
+        flushSegment('max_size')
+      }
+    },
+    (data, meta) => {
+      send(data, meta)
+    }
+  )
+
+  // Flush when the RUM view changes
+  const { unsubscribe: unsubscribeViewCreated } = lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, () => {
+    flushSegment('view_change')
+  })
+
+  // Flush when the session is renewed
+  const { unsubscribe: unsubscribeSessionRenewed } = lifeCycle.subscribe(LifeCycleEventType.SESSION_RENEWED, () => {
+    flushSegment('session_renewed')
+  })
+
+  // Flush when leaving the page
+  const { unsubscribe: unsubscribeBeforeUnload } = lifeCycle.subscribe(LifeCycleEventType.BEFORE_UNLOAD, () => {
+    flushSegment('before_unload')
+  })
+
+  // Flush when visibility changes
+  const { stop: unsubscribeVisibilityChange } = addEventListener(
+    emitter,
+    DOM_EVENT.VISIBILITY_CHANGE,
+    () => {
+      if (document.visibilityState === 'hidden') {
+        flushSegment('visibility_change')
```

Comment:
```suggestion
        flushSegment('page_hidden')
```
it would be more accurate

Reply:
This might be confusing if we support the `pagehide` event some day.

Reply:
something like `visibility_hidden` then? since change could either mean `visible` or `hidden`

Reply:
Ok!